dvc/command/metrics.py

```diff
@@ -14,31 +14,44 @@
 def _show_metrics(
     metrics, all_branches=False, all_tags=False, all_commits=False
 ):
-    from dvc.utils.diff import format_dict
+    from dvc.utils.diff import format_dict, table
     from dvc.utils.flatten import flatten
 
     # When `metrics` contains a `None` key, it means that some files
     # specified as `targets` in `repo.metrics.show` didn't contain any metrics.
     missing = metrics.pop(None, None)
 
-    lines = []
+    header_set = set()
+    rows = []
+    for _branch, val in metrics.items():
+        for _fname, metric in val.items():
+            for key, _val in flatten(format_dict(metric)).items():
+                header_set.add(key)
+    header = sorted(header_set)
     for branch, val in metrics.items():
-        if all_branches or all_tags or all_commits:
-            lines.append(f"{branch}:")
-
         for fname, metric in val.items():
+            row = []
+            if all_branches or all_tags or all_commits:
```

Comment:
Let's have 
```
with_rev = any(all_branches, all_tags, all_commits)
```
somewhere above and use `with_rev` here and in the `if` further below.